hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/DataSkippingUtils.scala

```diff
@@ -211,10 +211,10 @@ object DataSkippingUtils extends Logging {
           .map(colName => GreaterThan(genColNumNullsExpr(colName), Literal(0)))
 
       // Filter "colA is not null"
-      // Translates to "colA_nullCount = 0" for index lookup
+      // Translates to "colA_nullCount < colA_valueCount" for index lookup
       case IsNotNull(attribute: AttributeReference) =>
         getTargetIndexedColumnName(attribute, indexSchema)
-          .map(colName => EqualTo(genColNumNullsExpr(colName), Literal(0)))
+          .map(colName => LessThan(genColNumNullsExpr(colName), genColValueCountExpr))
```

Comment:
Filter `colA is not null` is the complement to `colA is null` then why the two have different translation (one has to depend on the valueCount while the other depends on Literal(0))? 

I mean if `colA is null` is translated to `GreaterThan(genColNumNullsExpr(colName), Literal(0))`, then shouldn't `colA is not null` be translated to `LessThanOrEqual(genColNumNullsExpr(colName), Literal(0))`?

Or if you say that `colA is not null` should be translated to `LessThan(genColNumNullsExpr(colName), genColValueCountExpr)`, then shouldn't `colA is null` be translated to `GreaterThanOrEqual(genColNumNullsExpr(colName), genColValueCountExpr)`?

Reply:
Ha! Great question. 

The reason is logical fallacy: "colA is not null" != "there is no null in colA", instead it's "colA contains non-null" (you can check out some other expressions, there are many expressions in this list that carry such properties):

 - "is null" means that the column has to contain null values (ie `nullCount` > 0)
 - "is not null" means that the column has to contain non-null values (ie `nullCount` < `valueCount`)